server.js

```diff
@@ -2056,14 +2047,57 @@ if (config.startServer) {
           process.exit(0);
         }
 
-        process.on('SIGTERM', () => {
-          opentelemetry
-            .shutdown()
-            .catch((err) => logger.error('Error shutting down OpenTelemetry', err))
-            .finally(() => {
-              process.exit(0);
-            });
-        });
+        const prepareForTermination = async () => {
+          // We want to proceed with termination even if something goes wrong,
+          // so don't allow this function to throw.
+          try {
+            await util.promisify(server.close).call(server);
+
+            await socketServer.close();
+
+            // Wait for all currently-executing cron jobs to finish.
+            await cron.stop();
+
+            // Wait for all currently-executing server jobs to finish.
+            await serverJobs.stop();
+          } catch (err) {
+            logger.error('Error while shutting down server', err);
+            Sentry.captureException(err);
+          }
+        };
+
+        const terminate = async () => {
+          // Shut down OpenTelemetry exporting.
+          try {
+            opentelemetry.shutdown();
+          } catch (err) {
+            logger.error('Error shutting down OpenTelemetry', err);
+            Sentry.captureException(err);
+          }
+
+          // Flush all events to Sentry.
+          try {
+            await Sentry.flush();
+          } finally {
+            process.exit(0);
+          }
+        };
+
+        // If we're running in EC2 auto scaling, this will wait for our instance
+        // to enter the `Terminating:Wait` state.
+        //
+        // If we're not running in EC2, this will be a no-op.
+        lifecycleHooks
+          .handleAndCompleteInstanceTermination(prepareForTermination)
+          .catch((err) => {
+            logger.error('Error handling instance termination', err);
+            Sentry.captureException(err);
+          })
+          .finally(terminate);
+
+        // Also listen for and handle SIGTERM, which can be sent to gracefully
+        // shutdown a process.
+        process.on('SIGTERM', () => prepareForTermination().finally(terminate));
```

Comment:
In theory, this will allow us to handle `systemctl stop`/`systemctl restart`, though we should validate that we've correctly configured it to send `SIGTERM`.